### PR TITLE
Add a new transpiler pass to remove all barriers

### PR DESCRIPTION
**Pull request by peendebak**
### Summary

Add pass to remove all barriers.

### Details and comments

Implementation is based on the `MergeAdjacentBarriers` pass.
A minimal example:
```
from qiskit import QuantumCircuit
from qiskit.transpiler import PassManager
from qiskit.transpiler.passes import RemoveBarriers

circuit = QuantumCircuit(1)
circuit.barrier()
circuit.x(0)
circuit.barrier()
circuit.h(0)
pm = PassManager(RemoveBarriers())
circuit = pm.run(circuit)
circuit.draw()
```

### CODE REVIEW

**Comment by peendebak** [comment]
@1ucian0 Could please you review this PR?

**Review by ajavadia — COMMENTED** [review]
Looks good to me, can you add a release note (using `reno`)?
I think in principle having this pass can be handy, but it is not semantics preserving as barriers can have semantic information. Maybe add a note to that effect as well.

**Comment by peendebak** [comment]
@ajavadia I updated the docstring to note the transformation does not preserve semantics. Can I describe this more precisely? Any transformation that is not the identity does not preserve semantics right? And the removal of barriers leaves the unitary intact, so those semantics are not changed.

I added release notes as well.

**Comment by 1ucian0** [comment]
the docstring seems to have an error.

**Comment by peendebak** [comment]
@1ucian0 The error message generated is not very helpful. It seems to be related to `plot_state_paulivec`, which is not related at all to this PR.

image

I cannot reproduce the error locally. Can you give suggestions on how to tackle this?

**Comment by ajavadia** [comment]
> Any transformation that is not the identity does not preserve semantics right?
Hmm I guess it depends on the definition of semantics. If you only consider the unitary then yes it's semantics-preserving. But I meant to say that barriers are not just sugar in the circuit; they mean something (for example they indicate some user's intent about two gates not merging). So by removing those you are removing that extra information on the circuit. 

I suppose you could argue that collapsing two back-to-back X gates into identity is also changing something about the circuit, for example the type of noise that occurs on it during execution. But the semantics of OpenQASM allow this, but does not allow barrier removal. I only accepted it here because it can be a useful utility for users, but I wanted that information included.

These types of "valid" compiler transformations that tie into the semantics of the QASM language are discussed a bit in the [OpenQASM 3 paper](https://arxiv.org/abs/2104.14722), especially in the delay and timing section.

**Review by ajavadia — APPROVED** [review]

**Comment by peendebak** [comment]
> > Any transformation that is not the identity does not preserve semantics right?
> 
> Hmm I guess it depends on the definition of semantics. If you only consider the unitary then yes it's semantics-preserving. But I meant to say that barriers are not just sugar in the circuit; they mean something (for example they indicate some user's intent about two gates not merging). So by removing those you are removing that extra information on the circuit.
> 
> I suppose you could argue that collapsing two back-to-back X gates into identity is also changing something about the circuit, for example the type of noise that occurs on it during execution. But the semantics of OpenQASM allow this, but does not allow barrier removal. I only accepted it here because it can be a useful utility for users, but I wanted that information included.
> 
> These types of "valid" compiler transformations that tie into the semantics of the QASM language are discussed a bit in the [OpenQASM 3 paper](https://arxiv.org/abs/2104.14722), especially in the delay and timing section.

@ajavadia Thanks for clarifying. I added type annotations and fixed a typo. PR should be remove for merge now.

**Review by ajavadia — DISMISSED** [review]

File: docs/apidocs/terra.rst [hunk 1]
```diff
@@ -33,6 +33,5 @@ Qiskit Terra API Reference
    transpiler_preset
    utils
    validation
-   visualization
```
Comment (1ucian0) [inline docs/apidocs/terra.rst:36]: Is this change related?

File: qiskit/transpiler/passes/utils/remove_barriers.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
Comment (1ucian0) [inline qiskit/transpiler/passes/utils/remove_barriers.py:3]:
```suggestion
# (C) Copyright IBM 2017, 2021.
```

File: qiskit/transpiler/passes/utils/remove_barriers.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        # add the merged barriers to a new DAG
+        new_dag = dag._copy_circuit_metadata()
+
+        # go over current nodes, and add them to the new dag
+        for node in dag.topological_op_nodes():
+            if node.name == "barrier":
+                pass
+            else:
+                # copy the condition over too
+                new_dag.apply_operation_back(node.op, qargs=node.qargs, cargs=node.cargs)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/utils/remove_barriers.py:55]: Why not `dag.remove_all_ops_named("barrier")`?

**Review by 1ucian0 — DISMISSED** [review]
I think `DAGCircuit.remove_all_ops_named()` is exactly for this kind of situations (even maybe better performance?)

File: qiskit/transpiler/passes/utils/remove_barriers.py (hunk 3 above)
Comment (eendebakpt) [reply qiskit/transpiler/passes/utils/remove_barriers.py:55]: @1ucian0 I as not aware of this method, but that is indeed better. I updated the code. 

One question: the method `dag.remove_all_ops_named` removes the operations in-place. Is that OK? I see that the `run()` method for many passes returns a new dag, but the argument dag is modified inplace as well. In the base class for passes, the run method is not annotated with a return value:

https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/transpiler/basepasses.py#L70

File: docs/apidocs/terra.rst (hunk 1 above)
Comment (eendebakpt) [reply docs/apidocs/terra.rst:36]: No, I will remove it.

File: test/python/transpiler/test_remove_barriers.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
```
Comment (1ucian0) [inline test/python/transpiler/test_remove_barriers.py:3]:
```suggestion
# (C) Copyright IBM 2017, 2021.
```

File: qiskit/transpiler/passes/utils/remove_barriers.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            from qiskit.transpiler import PassManager
+            from qiskit.transpiler.passes import RemoveBarriers
+
+            circuit = QuantumCircuit(1)
+            circuit.x(0)
+            circuit.barrier()
+            circuit.h(0)
+
+            pm = PassManager(RemoveBarriers())
+            circuit = pm.run(circuit)
```
Comment (1ucian0) [inline qiskit/transpiler/passes/utils/remove_barriers.py:38]: You can call a pass directly, no need to build a pass manager.
```suggestion
            from qiskit import QuantumCircuit
            from qiskit.transpiler.passes import RemoveBarriers

            circuit = QuantumCircuit(1)
            circuit.x(0)
            circuit.barrier()
            circuit.h(0)

            circuit = RemoveBarriers()(circuit)
```

File: qiskit/transpiler/passes/utils/remove_barriers.py (hunk 3 above)
Comment (1ucian0) [reply qiskit/transpiler/passes/utils/remove_barriers.py]: I dont think we have any warranty about input preservation. I was trying to play a bit with the pass and it seems like everything seems fine.

**Review by ajavadia — APPROVED** [review]

File: releasenotes/notes/add-remove-barriers-pass-dab69660f61e9d75.yaml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+---
+features:
+  - |
+    Add pass RemoveBarriers to remove all barriers in a circuit.
```
Comment (mtreinish) [inline releasenotes/notes/add-remove-barriers-pass-dab69660f61e9d75.yaml:4]:
```suggestion
    Added a new transpiler pass, :class:`~qiskit.transpiler.passes.RemoveBarriers`,
    to :mod:`qiskit.transpiler.passes`. This pass is used to remove all barriers in a
    circuit.
```

**Review by mtreinish — APPROVED** [review]

**Review by mtreinish — APPROVED** [review]
